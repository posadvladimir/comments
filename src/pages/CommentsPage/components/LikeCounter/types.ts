
import type { LikeIconStates } from '../../common/ts/enums/likeIconStates';

export type TLikeCounterProps = {
    likes: number;
    iconState?: LikeIconStates;
    disabled?: boolean;
}