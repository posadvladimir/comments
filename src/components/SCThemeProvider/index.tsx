
import { ThemeProvider } from 'styled-components';

import type { TChildrenProps } from 'src/ts/types/children';

import theme from 'src/styles/theme';

const SCThemeProvider: React.FC<TChildrenProps> = ({ children }) => (
    <ThemeProvider theme={theme}>
        {children}
    </ThemeProvider>
);

export default SCThemeProvider;