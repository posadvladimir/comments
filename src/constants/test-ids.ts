
const LOADER_TEST_ID = 'loader-test-id';
const BUTTON_TEST_ID = 'button-test-id';
const LIKE_COUNTER_TEST_ID = 'like-counter-test-id';
const LIKE_COUNTER_BUTTON_TEST_ID = 'like-counter-button-test-id';
const LIKE_COUNTER_TOTAL_TEST_ID = 'like-counter-total-test-id';
const COMMENT_TEST_ID = 'comment-test-id';
const ERROR_TEST_ID = 'error-test-id';

export {
    LOADER_TEST_ID,
    BUTTON_TEST_ID,
    LIKE_COUNTER_TEST_ID,
    LIKE_COUNTER_BUTTON_TEST_ID,
    COMMENT_TEST_ID,
    LIKE_COUNTER_TOTAL_TEST_ID,
    ERROR_TEST_ID,
};