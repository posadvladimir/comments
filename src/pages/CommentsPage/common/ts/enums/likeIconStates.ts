
export enum LikeIconStates {
    GreyStroke = 'grey-stroke',
    RedStroke = 'red-stroke',
    FullRed = 'full-red',
}