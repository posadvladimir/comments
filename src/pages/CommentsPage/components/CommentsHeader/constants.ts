
import type { TDeclensionArray } from '../../common/ts/types/declensions';

export const COMMENT_DECLENSIONS: TDeclensionArray = [
    'комментарий',
    'комментария',
    'комментариев'
];