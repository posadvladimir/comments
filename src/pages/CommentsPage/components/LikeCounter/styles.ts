
import { styled } from 'styled-components';

export const LikeCounterWrapper = styled.div`
    display: flex;
    line-height: 150%;
`;

export const LikeCounterButton = styled.button`
    cursor: pointer;

    &[disabled] {
        cursor: auto;
    }
`;

const LIKE_IMG_SIZE = 22;
const LIKE_IMG_SIZE_PHONE = 20;

export const LikeCounterButtonImg = styled.img`
    width: ${LIKE_IMG_SIZE}px;
    height: ${LIKE_IMG_SIZE}px;

    @media ${props => props.theme.media.phone} {
        width: ${LIKE_IMG_SIZE_PHONE}px;
        height: ${LIKE_IMG_SIZE_PHONE}px;
    }
`;

export const LikeCounterTotal = styled.span`
    margin-left: 8px;
    font-size: ${props => props.theme.default.fontSize - 1}px;
    font-weight: 700;

    @media ${props => props.theme.media.phone} {
        font-size: inherit;
    }
`;