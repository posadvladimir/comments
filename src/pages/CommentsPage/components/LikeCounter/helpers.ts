
import { LikeIconStates } from '../../common/ts/enums/likeIconStates';

import greyStrokeLikeIcon from 'src/assets/likes/grey-stroke.png';
import redStrokeLikeIcon from 'src/assets/likes/red-stroke.png';
import fullRedLikeIcon from 'src/assets/likes/full-red.png';

const likeIcons: Record<LikeIconStates, string> = {
    [LikeIconStates.GreyStroke]: greyStrokeLikeIcon,
    [LikeIconStates.RedStroke]: redStrokeLikeIcon,
    [LikeIconStates.FullRed]: fullRedLikeIcon,
};

export const getLikeIcon = (likeIcon: LikeIconStates): string => {
    return likeIcons[likeIcon];
};