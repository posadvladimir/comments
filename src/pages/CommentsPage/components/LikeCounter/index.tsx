
import { useState, useContext } from 'react';

import type { TLikeCounterProps } from './types';

import { CommentsPageContext } from '../../common/contexts/CommentsPageContext';

import { LikeIconStates } from '../../common/ts/enums/likeIconStates';

import {
    LIKE_COUNTER_TEST_ID,
    LIKE_COUNTER_BUTTON_TEST_ID,
    LIKE_COUNTER_TOTAL_TEST_ID,
} from 'src/constants/test-ids';

import { getRuNumberFormat } from '../../common/helpers/getRuNumberFormat';
import { getLikeIcon } from './helpers';

import * as S from './styles';

export default function LikeCounter({
    likes,
    iconState = LikeIconStates.RedStroke,
    disabled,
}: TLikeCounterProps) {
    const [isClicked, setIsClicked] = useState(false);
    const [isLiked, setIsLiked] = useState(false);

    const { increaseTotalLikes, decreaseTotalLikes } = useContext(CommentsPageContext);

    const handleLikeToggle = () => {
        if (!isClicked) {
            setIsClicked(true);
        }

        if (!isLiked) {
            setIsLiked(true);
            increaseTotalLikes();
        } else {
            setIsLiked(false);
            decreaseTotalLikes();
        }
    };

    let likesCount = likes;
    let likeIconState = iconState;

    if (isClicked && isLiked) {
        likesCount++;
        likeIconState = LikeIconStates.FullRed;
    }
    if (isClicked && !isLiked && (likeIconState === LikeIconStates.FullRed)) {
        likesCount--;
        likeIconState = LikeIconStates.RedStroke;
    }

    return (
        <S.LikeCounterWrapper data-testid={LIKE_COUNTER_TEST_ID}>
            <S.LikeCounterButton
                onClick={handleLikeToggle}
                disabled={disabled}
                data-testid={LIKE_COUNTER_BUTTON_TEST_ID}
            >
                <S.LikeCounterButtonImg
                    src={getLikeIcon(likeIconState)}
                    alt="like icon"
                />
            </S.LikeCounterButton>
            <S.LikeCounterTotal data-testid={LIKE_COUNTER_TOTAL_TEST_ID}>
                {getRuNumberFormat(likesCount)}
            </S.LikeCounterTotal>
        </S.LikeCounterWrapper>
    );
}