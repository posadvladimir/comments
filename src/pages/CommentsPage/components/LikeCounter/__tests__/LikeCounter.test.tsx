
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';

import { LIKE_COUNTER_BUTTON_TEST_ID, LIKE_COUNTER_TOTAL_TEST_ID } from 'src/constants/test-ids';

import { LikeIconStates } from 'src/pages/CommentsPage/common/ts/enums/likeIconStates';

import SCThemeProvider from 'src/components/SCThemeProvider';
import LikeCounter from '..';

const LIKES_TOTAL = 1234;

const iconTestData = [
    { iconState: LikeIconStates.GreyStroke, expectedSrc: 'grey-stroke.png' },
    { iconState: LikeIconStates.RedStroke, expectedSrc: 'red-stroke.png' },
    { iconState: LikeIconStates.FullRed, expectedSrc: 'full-red.png' },
];

const getlikeCounterComponent = (likes: number, iconState: LikeIconStates, disabled = false): JSX.Element => (
    <SCThemeProvider>
        <LikeCounter
            likes={likes}
            iconState={iconState}
            disabled={disabled}
        />
    </SCThemeProvider>
)

const renderLikeCounter = (likes: number, iconState: LikeIconStates, disabled = false) => {
    render(getlikeCounterComponent(likes, iconState, disabled));
};

describe('LikeCounter', () => {
    test.each(iconTestData)('отображается правильная иконка для %s', ({ iconState, expectedSrc }) => {
        renderLikeCounter(1, iconState);

        const likeImgElement = screen.getByAltText('like icon');

        expect(likeImgElement).toHaveAttribute('src', expectedSrc);
    });

    test('отображает количество лайков в правильном формате', () => {    
        renderLikeCounter(LIKES_TOTAL, LikeIconStates.GreyStroke);
    
        const formattedLikes = screen.getByText('1 234');
    
        expect(formattedLikes).toBeInTheDocument();
    });

    test('переключается иконка лайка при клике', () => {
        renderLikeCounter(LIKES_TOTAL, LikeIconStates.RedStroke);

        const likeButton = screen.getByTestId(LIKE_COUNTER_BUTTON_TEST_ID);
        const likeImgElement = screen.getByAltText('like icon');

        userEvent.click(likeButton);

        expect(likeImgElement).toHaveAttribute('src', 'full-red.png');
    });

    test('переключается иконка лайка при двойном клике', () => {
        renderLikeCounter(LIKES_TOTAL, LikeIconStates.RedStroke);

        const likeButton = screen.getByTestId(LIKE_COUNTER_BUTTON_TEST_ID);
        const likeImgElement = screen.getByAltText('like icon');

        userEvent.click(likeButton);
        userEvent.click(likeButton);

        expect(likeImgElement).toHaveAttribute('src', 'red-stroke.png');
    });

    test('не меняется иконка лайка при дизейбле', () => {
        renderLikeCounter(LIKES_TOTAL, LikeIconStates.GreyStroke, true);

        const likeButton = screen.getByTestId(LIKE_COUNTER_BUTTON_TEST_ID);
        const likeImgElement = screen.getByAltText('like icon');

        userEvent.click(likeButton);

        expect(likeImgElement).toHaveAttribute('src', 'grey-stroke.png');
    });

    test('увеличивается значение кол-ва лайков', () => {
        renderLikeCounter(LIKES_TOTAL, LikeIconStates.RedStroke);

        const likeButton = screen.getByTestId(LIKE_COUNTER_BUTTON_TEST_ID);
        const likeTotalElement = screen.getByTestId(LIKE_COUNTER_TOTAL_TEST_ID);

        userEvent.click(likeButton);

        expect(likeTotalElement).toHaveTextContent('1 235');
    });

    test('возвращается к исходному значению кол-во лайков при двойном клике', () => {
        renderLikeCounter(LIKES_TOTAL, LikeIconStates.RedStroke);

        const likeButton = screen.getByTestId(LIKE_COUNTER_BUTTON_TEST_ID);
        const likeTotalElement = screen.getByTestId(LIKE_COUNTER_TOTAL_TEST_ID);

        userEvent.click(likeButton);
        userEvent.click(likeButton);

        expect(likeTotalElement).toHaveTextContent('1 234');
    });

    test('не меняется значение кол-ва лайков при дизейбле', () => {
        renderLikeCounter(LIKES_TOTAL, LikeIconStates.GreyStroke, true);

        const likeButton = screen.getByTestId(LIKE_COUNTER_BUTTON_TEST_ID);
        const likeTotalElement = screen.getByTestId(LIKE_COUNTER_TOTAL_TEST_ID);

        userEvent.click(likeButton);

        expect(likeTotalElement).toHaveTextContent('1 234');
    });

    test.each(iconTestData)('соотвествует снепшоту', ({ iconState }) => {
        const tree = renderer.create(getlikeCounterComponent(1234, iconState)).toJSON();

        expect(tree).toMatchSnapshot();
    });
});