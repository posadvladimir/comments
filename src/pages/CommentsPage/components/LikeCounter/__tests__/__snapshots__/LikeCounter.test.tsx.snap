// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LikeCounter соотвествует снепшоту 1`] = `
<div
  className="sc-bcPKhP jPNrFD"
  data-testid="like-counter-test-id"
>
  <button
    className="sc-grXZZQ gdqHVB"
    data-testid="like-counter-button-test-id"
    disabled={false}
    onClick={[Function]}
  >
    <img
      alt="like icon"
      className="sc-dkjKgF gPDvA-D"
      src="grey-stroke.png"
    />
  </button>
  <span
    className="sc-hLclGa dtULLj"
    data-testid="like-counter-total-test-id"
  >
    1 234
  </span>
</div>
`;

exports[`LikeCounter соотвествует снепшоту 2`] = `
<div
  className="sc-bcPKhP jPNrFD"
  data-testid="like-counter-test-id"
>
  <button
    className="sc-grXZZQ gdqHVB"
    data-testid="like-counter-button-test-id"
    disabled={false}
    onClick={[Function]}
  >
    <img
      alt="like icon"
      className="sc-dkjKgF gPDvA-D"
      src="red-stroke.png"
    />
  </button>
  <span
    className="sc-hLclGa dtULLj"
    data-testid="like-counter-total-test-id"
  >
    1 234
  </span>
</div>
`;

exports[`LikeCounter соотвествует снепшоту 3`] = `
<div
  className="sc-bcPKhP jPNrFD"
  data-testid="like-counter-test-id"
>
  <button
    className="sc-grXZZQ gdqHVB"
    data-testid="like-counter-button-test-id"
    disabled={false}
    onClick={[Function]}
  >
    <img
      alt="like icon"
      className="sc-dkjKgF gPDvA-D"
      src="full-red.png"
    />
  </button>
  <span
    className="sc-hLclGa dtULLj"
    data-testid="like-counter-total-test-id"
  >
    1 234
  </span>
</div>
`;
